### PR TITLE
Get current window frame from Monitor Selection Service

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes https://github.com/dotnet/roslyn/issues/5698

### CODE REVIEW

**Comment by jmarolf** [comment]
@dotnet/roslyn-ide @amcasey Please Review

**Comment by jmarolf** [comment]
test prtest/win/dbg/eta please

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by mmitche** [comment]
@dotnet-bot test this please

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 1]
```diff
@@ -649,23 +649,34 @@ public override void CloseAdditionalDocument(DocumentId documentId)
...
             if (documentId == null)
             {
-                frame = null;
-                factory = null;
-                return false;
+                var monitorSelectionService = ServiceProvider.GetService(typeof(SVsShellMonitorSelection)) as IVsMonitorSelection;
+                object value = null;
+                if (monitorSelectionService != null &&
+                   (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out value)) ||
+                    ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_WindowFrame, out value))))
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Is it right that we prefer the document if possible?  There isn't a way to just get whatever is focused?

**Comment by Pilchie** [comment]
Would be good to have the *why* in the commit message (so that the info bar will show if invoked in a toolwindow like Interactive), for people spelunking history in the future.

**Comment by Pilchie** [comment]
:+1:

**Comment by dpoeschl** [comment]
Did the @mmitche bot go crazy?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 1 above)
Comment (dpoeschl) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: We really shouldn't use that ```documentId```. The scenario of a codefix/refactoring encountering an error when no documents are open is certainly the main driving scenario for this change, but if we're making this change in the no-document case then we should make the change consistently, for all cases. For example, if I have Main VS on Monitor 1 and Interactive on Monitor 2, and a codefix has a problem while I'm working in Interactive, I don't want to have to look over to Monitor 1 to see the infobar *sometimes* (iff there are no documents open over the entire VS process).

Also, does this work correctly in Peek? (I have no idea how Peek works.)

**Comment by mmitche** [comment]
Was debugging the ci issue.  Closing and reopening is a way of retriggering.

Sent from my Windows Phone
________________________________

**Comment by tannergooding** [comment]
test this please

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 2]
```diff
@@ -649,23 +649,34 @@ public override void CloseAdditionalDocument(DocumentId documentId)
...
         public bool TryGetInfoBarData(DocumentId documentId, out IVsWindowFrame frame, out IVsInfoBarUIFactory factory)
         {
+            frame = null;
+            factory = null;
             if (documentId == null)
             {
-                frame = null;
-                factory = null;
-                return false;
+                var monitorSelectionService = ServiceProvider.GetService(typeof(SVsShellMonitorSelection)) as IVsMonitorSelection;
```
Comment (amcasey) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Regular cast?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 3]
```diff
@@ -649,23 +649,34 @@ public override void CloseAdditionalDocument(DocumentId documentId)
...
+                object value = null;
+                if (monitorSelectionService != null &&
+                   (ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_DocumentFrame, out value)) ||
+                    ErrorHandler.Succeeded(monitorSelectionService.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_WindowFrame, out value))))
+                {
+                    frame = value as IVsWindowFrame;
```
Comment (amcasey) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Regular cast?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 4]
```diff
@@ -649,23 +649,34 @@ public override void CloseAdditionalDocument(DocumentId documentId)
...
+                {
+                    return false;
+                }
             }
 
-            frame = null;
-            factory = null;
-            return false;
+            factory = ServiceProvider.GetService(typeof(SVsInfoBarUIFactory)) as IVsInfoBarUIFactory;
+            return frame != null && factory != null;
```
Comment (amcasey) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs]: Can ```frame``` be null here?

**Comment by jmarolf** [comment]
retest prtest/lin/dbg/unit32  please

**Comment by jmarolf** [comment]
retest eta please

**Comment by jmarolf** [comment]
retest unit64 please

**Pull request merged**